pkg/frontend/admin_openshiftcluster_serialconsole.go

```diff
@@ -0,0 +1,156 @@
+	if err != nil {
+		return err
+	}
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return err
+	}
+
+	resource, err := azure.ParseResourceID(doc.OpenShiftCluster.ID)
+	if err != nil {
+		return err
+	}
+
+	fpAuthorizer, err := f.env.FPAuthorizer(
+		doc.OpenShiftCluster.Properties.ServicePrincipalProfile.TenantID, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
+
+	s, err := newSerialTool(log, f.computeClientFactory(resource.SubscriptionID, fpAuthorizer),
+		storage.NewAccountsClient(resource.SubscriptionID, fpAuthorizer), doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+
+	return s.dump(ctx, vmName, w)
+}
+
+type serialTool struct {
+	log *logrus.Entry
+	oc  *api.OpenShiftCluster
+
+	accounts        storage.AccountsClient
+	virtualMachines compute.VirtualMachinesClient
+
+	clusterResourceGroup string
+}
+
+func newSerialTool(log *logrus.Entry, virtualMachines compute.VirtualMachinesClient,
+	accounts storage.AccountsClient, oc *api.OpenShiftCluster) (*serialTool, error) {
+
+	return &serialTool{
+		log: log,
+		oc:  oc,
+
+		accounts:        accounts,
+		virtualMachines: virtualMachines,
+
+		clusterResourceGroup: stringutils.LastTokenByte(oc.Properties.ClusterProfile.ResourceGroupID, '/'),
+	}, nil
+}
+
+// dump writes to w the output of the serial console for vmname in raw bytes
+func (s *serialTool) dump(ctx context.Context, vmname string, w http.ResponseWriter) error {
+	vm, err := s.virtualMachines.Get(ctx, s.clusterResourceGroup, vmname, mgmtcompute.InstanceView)
+	if err != nil {
+		return err
+	}
+
+	u, err := url.Parse(*vm.InstanceView.BootDiagnostics.SerialConsoleLogBlobURI)
+	if err != nil {
+		return err
+	}
+
+	parts := strings.Split(u.Path, "/")
+	if len(parts) != 3 {
+		return fmt.Errorf("serialConsoleLogBlobURI has %d parts, expected 3", len(parts))
+	}
+
+	t := time.Now().UTC().Truncate(time.Second)
+	res, err := s.accounts.ListAccountSAS(ctx, s.clusterResourceGroup, "cluster"+s.oc.Properties.StorageSuffix, mgmtstorage.AccountSasParameters{
+		Services:               mgmtstorage.B,
+		ResourceTypes:          mgmtstorage.SignedResourceTypesO,
+		Permissions:            mgmtstorage.R,
+		Protocols:              mgmtstorage.HTTPS,
+		SharedAccessStartTime:  &date.Time{Time: t},
+		SharedAccessExpiryTime: &date.Time{Time: t.Add(24 * time.Hour)},
+	})
+	if err != nil {
+		return err
+	}
+
+	v, err := url.ParseQuery(*res.AccountSasToken)
+	if err != nil {
+		return err
+	}
+
+	blobService := azstorage.NewAccountSASClient("cluster"+s.oc.Properties.StorageSuffix, v, azure.PublicCloud).GetBlobService()
```

Comment:
Not sure if we actually care about this, but -- would we instead want to return a 302 to a SAS URI with https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#Blob.GetSASURI rather than proxying the content?

Reply:
Might be data implications. SAS URL can be accessed later from anywhere.